src/frontend/src/optimizer/plan_node/stream_project.rs

```diff
@@ -82,19 +82,21 @@ impl StreamProject {
         let mut watermark_derivations = vec![];
         let mut nondecreasing_exprs = vec![];
         let mut out_watermark_columns = FixedBitSet::with_capacity(core.exprs.len());
+        let mut out_monotonicity_map = MonotonicityMap::new();
         for (expr_idx, expr) in core.exprs.iter().enumerate() {
             use monotonicity_variants::*;
             match analyze_monotonicity(expr) {
-                Inherent(Constant) => {
-                    // XXX(rc): we can produce one watermark on each recovery for this case.
-                }
                 Inherent(monotonicity) => {
-                    if monotonicity.is_non_decreasing() {
+                    out_monotonicity_map.insert(expr_idx, monotonicity);
+                    if monotonicity.is_non_decreasing() && !monotonicity.is_constant() {
+                        // TODO(rc): may be we should also derive watermark for constant later
```

Comment:
According to the usage pattern of `monotonicity.is_non_decreasing() && !monotonicity.is_constant()`, it seems using EnumAsInner is more straightforward.

Reply:
Actually we should have derive watermark for constant here as well, while for not changing irrelevant logic, I kept it the same as before.

Reply:
IIUC, watermark should go up, but the constant watermark sounds trivial. Is there any use case in which we can benefit from a constant watermark?

Reply:
It's possible that we can benefit from getting a constant watermark for constant column each time we recover (if we don't derive watermark in this case, we won't know the watermark).